### PR TITLE
feat(toolbox): Support Toolbox 

### LINKED ISSUE #3104: Add toolbox container indicator
## Feature Request

[Toolbox](https://docs.fedoraproject.org/en-US/fedora-silverblue/toolbox/) has a feature where upon entering a container, a diamond is prefixed in the PS1 as a visual indicator that the user is working inside a container. `toolbox` will source [/etc/profile.d/toolbox.sh](https://github.com/containers/toolbox/blob/main/profile.d/toolbox.sh) automatically. This container indicator would be nice to have when using starship as well.

The script seems to check for the presence of `/run/.containerenv` and `/run/.toolboxenv`. Starship could check for the presence of the files to mimic the behavior, or somehow lean on `toolbox.sh` directly in case that functionality ever changes in the future.

### PR DESCRIPTION
**Pull request by alphacc**
#### Description
WIP to support toolbox from https://github.com/containers/toolbox

#### Motivation and Context
Have a better experience when using Fedora Silverblue

Fixes #3104

Heavily Inspired by the singularity module

I'd like to see if there is interest from other user, and I also would like to talk with toolbox upstream to see how we can better use env variable (TOOLBOX_ namespace ? etc...)

I just use the plugin daily for my own laptop.

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [ ] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

**Comment by matchai** [comment]
Hey Thomas 👋 

With less than 1000 GitHub stars, I'm afraid that Toolbox may be a little too obscure to qualify to be added to Starship.
Perhaps it's better suited as a [custom module](https://github.com/starship/starship/discussions/1252)?

**Comment by jankatins** [comment]
I've this as custom module (the checked files seem to be also used by the bash scripts to identify "in toolbox"): 

```toml
[custom.toolbox]
description = "The activated toolbox container"
command = "echo ${DISTTAG%container}"
when = """ [[ -f /run/.containerenv && -f /run/.toolboxenv ]] """
style = "bold red"
format = "[\\[Toolbox: $output\\]]($style) "
disabled = false
```

```bash
[22:01:31] [1] ✗  cat /etc/profile.d/toolbox.sh
[...]
if [ -f /run/.containerenv ] \
   && [ -f /run/.toolboxenv ]; then
    PS1=$(printf "\[\033[35m\]⬢\[\033[0m\]%s" "[\u@\h \W]\\$ ")
[...]
```

I would love to have this out of the box. We use this as $work to run our development env in there and fedora and GNOME seems to put a lot of effort into it, so my guess is that this will become biggish. A lot of news items on gnome 40 or fedora 34 mention it:

https://www.phoronix.com/scan.php?page=news_item&px=Fedora-Workstation-34-Excites
https://help.gnome.org/misc/release-notes/40.0/developers.html
https://www.theregister.com/2021/03/17/fedora_34_preview/

**Comment by vladimyr** [comment]
I second this. Toolbox and all the work on Fedora Silverblue have a high potential of producing the next big thing. It is definitely a niche but not obscure IMHO. :+1: from me.

**Comment by alphacc** [comment]
Ok thanks for the feedback I'll work on improving it and then we can decide.

One additional point is that FC34 will be the base for CentOS Stream 9 / RHEL 9 so it's likely it could gain more popularity soon.

**Comment by sclel016** [comment]
I'd definitely love to see this merge to Starship in some capacity.

File: src/configs/toolbox.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    pub format: &'a str,
+    pub style: &'a str,
+    pub disabled: bool,
+}
+
+impl<'a> Default for ToolboxConfig<'a> {
+    fn default() -> Self {
+        ToolboxConfig {
+            format: "[$symbol\\[$env\\]]($style) ",
+            symbol: "",
```
Comment (jankatins) [inline src/configs/toolbox.rs:18]: the bash one uses a red dot

**Comment by jkonecny12** [comment]
The toolbox repository is pretty close to 1000 :star:

**Comment by HarryMichal** [comment]
> The toolbox repository is pretty close to 1000 star

Already over :)

File: src/modules/toolbox.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+use super::{Context, Module, RootModuleConfig};
+
+use crate::configs::toolbox::ToolboxConfig;
+use crate::formatter::StringFormatter;
+
+/// Creates a module with the current toolbox image
+///
+/// Will display the toolbox image if `$toolbox_NAME` is set.
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let toolbox_env = context.get_env("TOOLBOX_PATH")?;
```
Comment (HarryMichal) [inline src/modules/toolbox.rs:10]: While `TOOLBOX_PATH` is set in every toolbx container, it does not hold information important for this particular module. It is a path to the internal Toolbx binary used for container initialization. Instead I'd look for the file `/run/.toolboxenv`. It's presence signifies that the container is in fact a toolbx container.

File: src/modules/toolbox.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    let toolbox_name = context.get_env("NAME")?;
```
Comment (HarryMichal) [inline src/modules/toolbox.rs:11]: Is set in our [Containerfile](https://github.com/containers/toolbox/blob/cfffb72fb0c275bfe32318c044a408603ca769f9/images/fedora/f36/Containerfile#L3) but we already have [reports](https://github.com/containers/toolbox/issues/188) of these variables interfering with other tools. While it is not possible to not use them for the foreseeable future due to the way Fedora OCI image machinery works, it is only specific to the Fedora images and they are not Toolbx-specific.

To read the name of the container, I'd read the file `/run/.containerenv` created by Podman (which is a dependency of Toolbx) and get the name from there. The content of the file can look like this:

```
engine="podman-3.4.2"
name="toolbox-dev"
id="236e7a414947759564bb2bc2d5cb3fc4a5999c7e8457bc763aaf5361a3f700c8"
image="registry.fedoraproject.org/fedora-toolbox:36"
imageid="3ef9a4eaf734b79fcafc0f846efb8a8e8850cfc81c66a9e8aea10c6637623a04"
rootless=1
graphRootMounted=1⏎
```

**Review by HarryMichal — CHANGES_REQUESTED** [review]
First of all, thank you for your work on this! I started using starship recently and it makes me happy that I don't have to add support for our tool to it myself.

I left some comments regarding the implementation. It mainly touches on the use of the different env vars. Also note that Toolbox has been [renamed](https://debarshiray.wordpress.com/2021/11/10/toolbox-is-now-toolbx/) to Toolbx (despite the fact that the rename has not occured yet in the implementation and most of the docs).

If you have any further questions, ping me :).

File: src/modules/toolbox.rs (hunk 2 above)
Comment (alphacc) [reply src/modules/toolbox.rs:10]: Thanks, yes it's better. I'll work on that.

File: src/modules/toolbox.rs (hunk 3 above)
Comment (alphacc) [reply src/modules/toolbox.rs:11]: noted, it's more elegant indeed.

File: src/configs/toolbox.rs (hunk 1 above)
Comment (alphacc) [reply src/configs/toolbox.rs:18]: Ok I'll add that !

**Comment by jankatins** [comment]
Another more generic "container" PR: https://github.com/starship/starship/pull/3304

**Pull request closed without merging**